pkg/util/connectgrpc/connectgrpc.go

```diff
@@ -115,17 +115,31 @@ func decodeRequest[Req any](req *httpgrpc.HTTPRequest) (*connect.Request[Req], e
 	return result, nil
 }
 
-func encodeRequest[Req any](req *connect.Request[Req]) (*httpgrpc.HTTPRequest, error) {
-	if req.Spec().Procedure == "" {
-		return nil, errors.New("cannot encode a request with empty procedure")
+type connectURLCtxKey struct{}
+
+func WithProcedure(ctx context.Context, u string) context.Context {
+	return context.WithValue(ctx, connectURLCtxKey{}, u)
+}
+
+func ProcedureFromContext(ctx context.Context) string {
+	s, _ := ctx.Value(connectURLCtxKey{}).(string)
+	return s
+}
+
+func encodeRequest[Req any](ctx context.Context, req *connect.Request[Req]) (*httpgrpc.HTTPRequest, error) {
+	url := req.Spec().Procedure
+	if url == "" {
+		if url = ProcedureFromContext(ctx); url == "" {
+			return nil, errors.New("cannot encode a request with empty procedure")
+		}
 	}
 	// The original Content-* headers could be invalidated,
 	// e.g. initial Content-Type could be 'application/json'.
 	h := removeContentHeaders(req.Header().Clone())
 	h.Set("Content-Type", "application/proto")
 	out := &httpgrpc.HTTPRequest{
 		Method:  http.MethodPost,
-		Url:     req.Spec().Procedure,
+		Url:     url,
 		Headers: connectHeaderToHTTPGRPCHeader(h),
 	}
 	var err error
```

Comment:
I don't really like it but I doubt there is a better way to ensure that the `Procedure` member of the `*connect.Request` is always present. If the request comes from the inside (i.e. not from the connect-generated handler), the caller does not have an option to specify it.

Reply:
Yes I think I hit that problem before, I think it is a reasonable approach.